### DISCUSSION TITLE
app.OnRecordBeforeCreateRequest() & Non-Empty Fields

### DISCUSSION THREAD

**Discussion by embowles93** [pocketbase/pocketbase#2881] [Q&A] [answered]
Hi there!

I've decided to be brave and make a foray in Go and the ability to extend the functionality of the tool. I'm working on an app for logging sightings of different types of planes. The additional functionality I am implementing is to have PocketBase make an http request to an external API to fetch relevant details about an airplane based on its registration, save that data and link it to a record as it is created. I've gotten the following code working successfully:

```
app.OnRecordBeforeCreateRequest().Add(func(e *core.RecordCreateEvent) error {
		// If the record belongs to the "sightings" collection
		reg:=e.Record.GetString("reg")

		if e.Record.Collection().Name == "sightings" {

			//logic to get some relevant data and create a new record with it, contained in the CreateAircraftRecord function.
			aircraftRecord, err := rapidapi.CreateAircraftRecord(app, reg)
			if err != nil {
				return err
			}

			// Set the "aircraft" field of the sighting record to the ID of the newly created aircraft record
			e.Record.Set("aircraft", aircraftRecord.Get("id"))
		}
		return nil
})
```

My question is, is it possible to implement this code **and** make the 'aircraft' relational field be 'nonempty'? I'd like to enforce this in the database if possible, but currently receive an error that that field is empty when I try and mark it as such. No worries if not - much appreciated for any help as I learn a new language!

**Comment by ganigeorgiev** [comment, marked as answer]
This eventually may change in the future as it was requested by other users too but at the moment the `OnRecordBeforeCreateRequest` hooks run after the schema fields validation.
Meaning that if you want to allow empty submitted values (_no matter whether you'll intercept the request processing and set the value on your own afterwards_) you'll have to remove the "Nonempty" check and overwrite the field value in the hook. A slightly simplified version of the above could be:

```go
app.OnRecordBeforeCreateRequest("sightings").Add(func(e *core.RecordCreateEvent) error {
    admin, _ := c.Get(apis.ContextAdminKey).(*models.Admin)
    if admin != nil {
        return nil // allow admins to submit anything
    }

    // logic to get some relevant data and create a new record with it, contained in the CreateAircraftRecord function.
    aircraftRecord, err := rapidapi.CreateAircraftRecord(app, e.Record.GetString("reg"))
    if err != nil {
        return err
    }

    // Set the "aircraft" field of the sighting record to the ID of the newly created aircraft record
    e.Record.Set("aircraft", aircraftRecord.Get("id"))
    
    return nil
})
```

To enforce the "Nonempty" check manually, you can also attach to the `OnRecordBeforeUpdateRequest` hook and return an error for example in case an empty record value was submitted.

**Reply by embowles93** [reply]
Many thanks @ganigeorgiev for the quick response! using the OnRecordBeforeUpdateRequest hook is a great idea for additional validation of a pseudo "non empty" check - I will try that. And appreciate the simplified code - I had not noticed I could put the collection name in the create hook to avoid an unnecessary if check 😀
